src/linker/Linker/LinkContext.cs

```diff
@@ -613,6 +614,20 @@ public void LogError (string text, int code, string subcategory = MessageSubCate
 			LogMessage (error);
 		}
 
+		/// <summary>
+		/// Display an error message to the end user.
+		/// </summary>
+		/// <param name="id">Unique error ID. Please see https://github.com/dotnet/linker/blob/main/docs/error-codes.md and https://github.com/dotnet/linker/blob/main/src/ILLink.Shared/DiagnosticId.cs for the list of errors and possibly add a new one</param>
+		/// <param name="subcategory">Optionally, further categorize this error</param>
+		/// <param name="origin">Filename, line, and column where the error was found</param>
+		/// <param name="args">Additional arguments to form a humanly readable message describing the warning</param>
+		/// <returns>New MessageContainer of 'Error' category</returns>
+		public void LogError (DiagnosticId id, string subcategory = MessageSubCategory.None, MessageOrigin? origin = null, params string[] args)
```

Comment:
I think this is probably good enough as a first try. I think we could go one step further:
* Make the origin parameter non-optional (just nullable). The fact that we don't specify origin in lot of callsites is effectively a bug - it makes the error much harder to fix.
* subcategory - I don't know if we even need it - at least in the "Simple" overload. As far as I can tell we almost never use it. So we could remove it from this overload completely.

And so we could change this to look like:
`LogError(MessageOrigin?, DiagnosticId, params string[] args)`

That way the callsites would not need to allocate string array manually. It does reorder the parameters, but personally I actually like the fact that it puts the origin first - it mimics how the error is presented to the user: "location: Id - message".

Reply:
I can implement those changes in this PR, the only question I have is: what about the other overloads for Logging errors? should we reorder the parameters too?

Reply:
I'd not change other overloads for now, only ones that take DiagnosticId

Reply:
I agree with Marek - base don a quick scan almost all callsites will go through the new overload, so the existing overloads can be considered "expert level" and I would not change them for now. Eventually we could even remove it in favor of the low-level approach of creating the `MessageContainer` manually at the callsite.

I assume we'll make similar changes to `LogWarning` in the other PR, right?

Reply:
> I assume we'll make similar changes to `LogWarning` in the other PR, right?

Yes, except that I'm not sure what's going to happen with the subcategory for LogWarning since there is needed

Reply:
In that case I would put is as a second parameter (origin, subcategory, id, args)...

In fact - can we deduce the subcategory from the id? I think we should be able to, right?

Reply:
We could create ranges and assign them a subcategory, but that would be flaky because we can add new diagnostics that don't fit the range.  Either way, there is an ilOffset optional argument that cannot be deduced in any way and is almost always null. Considering that optional arguments must appear after required arguments and that params need to be the last argument there will always be problems. I don't see how to not have to use the named arguments on this one.

Reply:
Since ilOffset is not of type `string` adding another overload (instead of optional parameter) should work - the overload resolution should correctly resolve it (since the types are not ambiguous).